Add transformations for placeholders

go2sh opened the pull request:
This PR adds transformations for placeholders by extending the syntax in the same way transformations are use with variables. Transformations are executed, when the user changes to the next placeholder.

I don't know if there is a better/more efficient way, but for me it works.

Fixes #34683.
TODO:
- [x] Add tests
- [x] Update documentation

PS: My first contribution to vscode. Great editor. :)
Best Regards,
go2sh

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=51621) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=51621) before we can accept your contribution.<br/><br/>:x: go2sh [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=51621)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode?pullRequest=51621) it.</sub>

go2sh commented:
CC @jrieken

menendezau commented:
Nice work!

jrieken commented:
👏 looks good so far, keep going esp. with adding tests. Check out `snippetParser.test.ts` and `snippetSession.test.ts`. Left a few questions about the grammar/rule-of-transform as comments

jrieken commented on src/vs/editor/contrib/snippet/snippetParser.ts:
```diff
@@ -722,13 +741,29 @@ export class SnippetParser {
+						if (this._parseTransform(placeholder)) {
+							parent.appendChild(placeholder);
+							return true;
+						}
```
Not sure if that's actually supported by TextMate. Do you have a reference?

go2sh replied:
As far as I know, TextMate only supports transformations for the simple case `${1/.../.../...}`. This two here would be an extension to the TextMate behavior. I agree, that the choice case might not be needed, but the nested one would be nice for default values e.g.

go2sh replied:
For Reference: http://manual.macromates.com/en/snippets#transformations

go2sh commented:
Another thing that differs from the TextMate behavior is that only the first placeholder is used for the transformation. I'll change that as well, so that every placeholder can holde it's own transformation.

go2sh commented:
@jrieken I fixed the code and added quite some tests. I hope this is enough for merging. ;-) 

Now VS Code supports:
```
${1/.../.../...}
${1:.../.../.../...}
${1|...|/.../.../...}
```
Every placeholder has its one transformation, so you can create two different results for the same index.

jrieken commented:
Thanks - quite busy the next few days but my hope is to squeeze this in during this week...

jrieken commented:
Bang on. Thanks, two things left

* please add a test in `snippetSession.test` that uses two (or more) selection at different indentation. Like `test('snippets, transform'...` but with multiple selection. I test this manually but it would be best to have tested too.
* update `snippet.md` with a placeholder transform section and please outline what isn't standard TM. 

Scheduled this now for June, let's ship this.

go2sh commented:
@jrieken Done. :)

jrieken commented:
Thanks.

jrieken commented:
Merged! 🍻 @go2sh

beausmith commented:
How to use `/upcase`, `/downcase/`, and `/capitalize`?

I've read the docs and tried:

```
"test": {
    "prefix": "test",
    "body": "${1} -> ${1:/upcase} ${1:/downcase}"
},
```

and

```
"test": {
    "prefix": "test",
    "body": "${1} -> ${1/upcase} ${1/downcase}"
},
```

…but neither work.

I don't see any tests for these transforms in this PR.

I'm stumped. I posted a question here:
https://stackoverflow.com/questions/51272365/vs-code-how-to-convert-snippet-placeholder-to-uppercase-or-lowercase

cc @go2sh

moranje commented:
The integer in the EBNF docs refer to a RegExp group not to a tabstop reference so:

```js
"test": {
    "prefix": "test",
    "body": "${1} -> ${1/(test)|(.)/${1:upcase}/}"
},
```

csaar95 commented:
In addition to #34683, it also doesn't support choice: `"${1|hello,world|} -> ${1/hello/hallo/}"`

thorn0 commented:
Can't make `${1:.../.../.../...}` work in 1.27.0-insiders.

go2sh commented:
@beausmith @moranje @csaar95 @thorn0 The vscode team removed all the advanced options as they create some regressions on existing snippets. See the docs. Only ${1/../../} is supported. Sorry for that. The removal also broke my usecase...

go2sh commented:
The grammar is clear. Transform is only valid for a tab-stop or a variable. The docs are not perfect. I try to make a pr. The Placeholder Transform section is from a time, where every thing was possible.

thorn0 commented:
Ah, now I see it.

tecosaur commented:
I'm trying a simple `"$1 ${1/(.*)/${1:/upcase}/}"` snippet but it isn't working for me (v1.40.0), has placeholder transform functionality just been completely removed?

mib32 commented:
> I'm trying a simple `"$1 ${1/(.*)/${1:/upcase}/}"` snippet but it isn't working for me (v1.40.0), has placeholder transform functionality just been completely removed?

Was struggling with this for a while, then understood - the transformation is not applied in real time - you have to hit Tab, and then it will get applied. Funny!
